blog/2020/12/package_latency.md

```diff
@@ -0,0 +1,718 @@
+
+This post introduces some new tools to detect the major sources of compiler latency, focusing specifically on the *type inference* step of compilation. [Type inference] is the process of deducing the types of all the internal variables in a method, given the types of the input arguments.  This is an early, but crucial, step in compilation, and the first to introduce substantial latency.  Later steps (optimization and native-code generation) are typically even more expensive, but in many cases inference latency is roughly proportional to these other forms of latency: typically, each run of inference is followed by these later steps of compilation.  (An exception are methods annotated with [`@nospecialize`], which get compiled to native code only for generic argument types.  Julia also has a number of heuristics which in essence add `@nospecialize` automatically for certain combinations of argument types.)
+
+In this post, we'll walk through the process of analyzing and optimizing the [DataFrames] package.  We chose DataFrames for several reasons:
+
+- DataFrames is widely used
+- the DataFrames API seems fairly stable, and they are approaching their 1.0 release
+- DataFrames is designed to work with any type of data and any user-supplied transformation functions, and so [faces substantial latency challenges](https://discourse.julialang.org/t/best-practices-for-profiling-compilation/48249)
+- DataFrames is developed by a sophisticated and conscientious team, and the package has already been [aggressively optimized for latency](https://discourse.julialang.org/t/release-announcements-for-dataframes-jl/18258/112?u=tim.holy) using tools that were, until now, state-of-the-art; this sets a high bar for any new tools (don't worry, we're going to crest that bar ;-) )
+- In a previous [blog post][invalidations], one of the authors indirectly "called out" DataFrames (and more accurately its dependency [CategoricalArrays]) for having a lot of difficult-to-fix invalidations.  To their credit, the developers made changes that dropped the number of invalidations by about 10×. This post is partly an attempt to return the favor.  That said, we hope they don't mind being guinea pigs for these new tools.
+
+This post is based on DataFrames 0.22.1, and version 0.9 of the underlying CategoricalArrays.  If you follow the steps of this blog post with different versions, you're likely to get different results from those shown here, partly because many of the issues we identified have been fixed in more recent releases.  It should also be emphasize that these analysis tools are only supported on Julia 1.6 and above; at the time of this post, Julia 1.6 not yet to "alpha" release phase but can be obtained from [nightly] snapshots or built from [source].
```

Comment:
"It should also be **emphasize** that" -> "It should also be emphasized that" I think? :)